### PR TITLE
[Capture] Prepare grad prim for catalyst translation

### PR DESCRIPTION
**Pull request by albi3ro**
**Context:**

When lowering to mlir, catalyst needs to know name and hash of the original function to properly name the `FuncOp` and to cache the created `FuncOp`, respectively.

See https://github.com/PennyLaneAI/catalyst/pull/2078 for the corresponding catalyst PR.

So we need to pass on the `fn` to the lowering.

**Description of the Change:**

* Stores `fn` as a piece of metadata in `grad_p`.

* Removed `jacobian_p` in favor of a single `grad_p` with a `scalar_out` kwarg. Less code to maintain.

**Benefits:**

We can define lowerings for the primitives in catalyst.

**Possible Drawbacks:**

**Related GitHub Issues:**

[sc-100398]

### CODE REVIEW

File: doc/releases/changelog-dev.md [hunk 1]
```diff
@@ -1133,6 +1133,8 @@
 
 <h3>Internal changes ⚙️</h3>
 
+* The `grad` and `jacobian` primitives now store the function under `fn`.
```
Comment (albi3ro) [inline doc/releases/changelog-dev.md]:
```suggestion
* The `grad` and `jacobian` primitives now store the function under `fn`.
  [(#8357)](https://github.com/PennyLaneAI/pennylane/pull/8357)
```

File: doc/releases/changelog-dev.md [hunk 2]
```diff
@@ -148,6 +148,9 @@
...
```
Comment (albi3ro) [inline doc/releases/changelog-dev.md]:
```suggestion
* The `grad` and `jacobian` primitives now store the function under `fn`. There is also now a single `grad_p`
   primitive for use in program capture.
```

File: doc/releases/changelog-dev.md [hunk 3]
```diff
@@ -148,6 +148,10 @@
 
 <h3>Internal changes ⚙️</h3>
 
+* The `grad` and `jacobian` primitives now store the function under `fn`. There is also now a single `grad_p`
+   primitive for use in program capture.
```
Comment (albi3ro) [inline doc/releases/changelog-dev.md]:
```suggestion
  primitive for use in program capture.
```

File: pennylane/_grad.py [hunk 4]
```diff
@@ -50,72 +53,43 @@ def _get_grad_prim():
...
+    def _grad_impl(*args, argnums, jaxpr, n_consts, method, h, scalar_out, fn):
+        if method != "auto":  # pragma: no cover
+            raise ValueError(f"Invalid values '{method=}' without QJIT.")
         consts = args[:n_consts]
         args = args[n_consts:]
 
         def func(*inner_args):
-            return jax.core.eval_jaxpr(jaxpr, consts, *inner_args)[0]
+            res = jax.core.eval_jaxpr(jaxpr, consts, *inner_args)
+            return res[0] if scalar_out else res
```
Comment (mudit2812) [inline pennylane/_grad.py:64]: Since `grad_prim.multiple_results = True`, shouldn't we always return `res` without unwrapping?

File: pennylane/_grad.py [hunk 5]
```diff
@@ -266,6 +256,14 @@ def __init__(self, func, argnums=None, h=None, method=None, *, argnum=None):
             # pylint:disable=unexpected-keyword-arg,no-value-for-parameter
             self._grad_fn = self._grad_with_forward(func, argnum=self._argnums)
 
+        # need to preserve input siganture for use in catalyst AOT compilation, but
+        # get rid of return annotation to placate autograd
+        self.__signature__ = inspect.signature(self._func).replace(
+            return_annotation=inspect.Signature.empty
+        )
+        n = getattr(self._func, "__name__", repr(self._func))
+        self.__name__ = f"<jacobian: {n}>"
```
Comment (mudit2812) [inline pennylane/_grad.py]: Shouldn't this say grad?
```suggestion
        self.__name__ = f"<grad: {n}>"
```

Comment (albi3ro) [reply pennylane/_grad.py]: This one is still the grad. jacobian is below.

File: pennylane/_grad.py (hunk 4 above)
Comment (albi3ro) [reply pennylane/_grad.py:64]: Ah... right. I'll add a `tree_leaves` call to the `scalar_out` branch too.

File: pennylane/_grad.py [hunk 6]
```diff
@@ -50,72 +53,43 @@ def _get_grad_prim():
...
 
         def func(*inner_args):
-            return jax.core.eval_jaxpr(jaxpr, consts, *inner_args)[0]
+            res = jax.core.eval_jaxpr(jaxpr, consts, *inner_args)
+            return res[0] if scalar_out else res
 
-        return jax.grad(func, argnums=argnums)(*args)
+        if scalar_out:
+            return jax.grad(func, argnums=argnums)(*args)
+        return jax.tree_util.tree_leaves(jax.jacobian(func, argnums=argnums)(*args))
```
Comment (albi3ro) [inline pennylane/_grad.py]:
```suggestion
           res = jax.grad(func, argnums=argnums)(*args)
        else:
           jax.jacobian(func, argnums=argnums)(*args)
        return jax.tree_util.tree_leaves(res)
```

File: pennylane/_grad.py [hunk 7]
```diff
@@ -50,72 +53,45 @@ def _get_grad_prim():
...
         def func(*inner_args):
-            return jax.core.eval_jaxpr(jaxpr, consts, *inner_args)[0]
+            res = jax.core.eval_jaxpr(jaxpr, consts, *inner_args)
+            return res[0] if scalar_out else res
 
-        return jax.grad(func, argnums=argnums)(*args)
+        if scalar_out:
+           res = jax.grad(func, argnums=argnums)(*args)
+        else:
+           jax.jacobian(func, argnums=argnums)(*args)
```
Comment (albi3ro) [inline pennylane/_grad.py]:
```suggestion
           res = jax.jacobian(func, argnums=argnums)(*args)
```

File: pennylane/_grad.py (hunk 5 above)
Comment (mudit2812) [reply pennylane/_grad.py]: Yeah, since jacobian is later in the file, shouldn't this one be grad? I'm still confused 😅

Comment (albi3ro) [reply pennylane/_grad.py]: I really have no idea what my brain was doing there...

File: pennylane/_grad.py (hunk 4 above)
Comment (albi3ro) [reply pennylane/_grad.py:64]: turns out it didn't matter since we always pass `argnums` as a list, so `jax.grad` will always return a list, even for a single input.

Comment (mudit2812) [reply pennylane/_grad.py:64]: But since you're unwrapping `res` when `scalar_out==True`, is that not going to break? When `multiple_results==True`, the primitive should return a list, right? Or is `res[0]` a list with a single scalar value?

Comment (albi3ro) [reply pennylane/_grad.py:64]: We're returning the gradient of that function.  And since `argnums` is always a list, the gradient will always be a list.

Comment (mudit2812) [reply pennylane/_grad.py:64]: Sorry about the confusion, I was being dumb

File: doc/releases/changelog-dev.md [hunk 8]
```diff
@@ -155,6 +155,10 @@
 
 <h3>Internal changes ⚙️</h3>
 
+* The `grad` and `jacobian` primitives now store the function under `fn`. There is also now a single `grad_p`
+  primitive for use in program capture.
+  [(#8357)](https://github.com/PennyLaneAI/pennylane/pull/8357)
```
Comment (dwierichs) [inline doc/releases/changelog-dev.md]: Wouldn't it make more sense to have the single primitive be called `jacobian_p`? 
As far as I know the naming, "gradient" implies scalar outputs of the original function.

Comment (dwierichs) [reply doc/releases/changelog-dev.md]: Very nice to see the reductions throughout for removing one of them, though 🎉

File: pennylane/_grad.py [hunk 9]
```diff
@@ -50,72 +53,45 @@ def _get_grad_prim():
     grad_prim.prim_type = "higher_order"
 
     @grad_prim.def_impl
-    def _grad_def_impl(*args, argnums, jaxpr, n_consts, method, h):
-        if method or h:  # pragma: no cover
-            raise ValueError(f"Invalid values '{method=}' and '{h=}' without QJIT.")
+    def _grad_impl(*args, argnums, jaxpr, n_consts, method, h, scalar_out, fn):
+        if method != "auto":  # pragma: no cover
+            raise ValueError(f"Invalid values '{method=}' without QJIT.")
```
Comment (dwierichs) [inline pennylane/_grad.py]:
```suggestion
            raise ValueError(f"Invalid value '{method=}' without QJIT.")
```

File: pennylane/_grad.py [hunk 10]
```diff
@@ -50,72 +53,45 @@ def _get_grad_prim():
...
+        else:
+            res = jax.jacobian(func, argnums=argnums)(*args)
+        return jax.tree_util.tree_leaves(res)
 
     # pylint: disable=unused-argument
     @grad_prim.def_abstract_eval
-    def _grad_abstract_eval(*args, argnums, jaxpr, n_consts, method, h):
-        if len(jaxpr.outvars) != 1 or jaxpr.outvars[0].aval.shape != ():
+    def _grad_abstract(*args, argnums, jaxpr, n_consts, method, h, scalar_out, fn):
+        if scalar_out and (len(jaxpr.outvars) != 1 or jaxpr.outvars[0].aval.shape != ()):
```
Comment (dwierichs) [inline pennylane/_grad.py]: I think this is closer to the underlying reasoning. It's completely equivalent, so really just subjective suggestion.
```suggestion
        if scalar_out and not (len(jaxpr.outvars) == 1 and jaxpr.outvars[0].aval.shape == ()):
```

File: pennylane/_grad.py [hunk 11]
```diff
@@ -266,6 +258,14 @@ def __init__(self, func, argnums=None, h=None, method=None, *, argnum=None):
             # pylint:disable=unexpected-keyword-arg,no-value-for-parameter
             self._grad_fn = self._grad_with_forward(func, argnum=self._argnums)
 
+        # need to preserve input siganture for use in catalyst AOT compilation, but
+        # get rid of return annotation to placate autograd
+        self.__signature__ = inspect.signature(self._func).replace(
+            return_annotation=inspect.Signature.empty
+        )
+        n = getattr(self._func, "__name__", repr(self._func))
+        self.__name__ = f"<grad: {n}>"
```
Comment (dwierichs) [inline pennylane/_grad.py]:
```suggestion
        fn_name = getattr(self._func, "__name__", repr(self._func))
        self.__name__ = f"<grad: {fn_name}>"
```

File: pennylane/_grad.py [hunk 12]
```diff
@@ -623,6 +623,14 @@ def __init__(self, func, argnums=None, method=None, h=None, *, argnum=None):
         self._method = method
         self._h = h
 
+        # need to preserve input siganture for use in catalyst AOT compilation, but
+        # get rid of return annotation to placate autograd
+        self.__signature__ = inspect.signature(self._func).replace(
+            return_annotation=inspect.Signature.empty
+        )
+        n = getattr(self._func, "__name__", repr(self._func))
+        self.__name__ = f"<jacobian: {n}>"
```
Comment (dwierichs) [inline pennylane/_grad.py]:
```suggestion
        fn_name = getattr(self._func, "__name__", repr(self._func))
        self.__name__ = f"<jacobian: {fn_name}>"
```

File: tests/capture/transforms/test_capture_cancel_inverses.py [hunk 13]
```diff
@@ -564,7 +563,7 @@ def f(x):
         jaxpr = jax.make_jaxpr(f)(1.5)
         assert len(jaxpr.eqns) == 3
         assert jaxpr.eqns[0].primitive == qml.RX._primitive
-        assert jaxpr.eqns[1].primitive == grad_prim if grad_fn == qml.grad else jacobian_prim
+        assert jaxpr.eqns[1].primitive == grad_prim
```
Comment (dwierichs) [inline tests/capture/transforms/test_capture_cancel_inverses.py]: Could check the `scalar_out` parameter 🤔 Non-blocking, though.

File: tests/capture/test_capture_diff.py [hunk 14]
```diff
@@ -24,51 +24,56 @@
...
-        def func(x):
-            return diff(jnp.sin, **kwargs)(x)
-
-        method = kwargs.get("method", None)
-        h = kwargs.get("h", None)
-        jaxpr = jax.make_jaxpr(func)(0.6)
-        with pytest.raises(ValueError, match=f"'{method=}' and '{h=}' without QJIT"):
-            func(0.6)
-        with pytest.raises(ValueError, match=f"'{method=}' and '{h=}' without QJIT"):
-            jax.core.eval_jaxpr(jaxpr.jaxpr, jaxpr.consts, 0.6)
```
Comment (dwierichs) [inline tests/capture/test_capture_diff.py:51]: Isn't there still the validation that `method` is not being passed on to plxpr? 🤔

File: tests/capture/test_capture_diff.py [hunk 15]
```diff
@@ -537,7 +543,10 @@ def circuit(x):
 
         jac_eqn = jaxpr.eqns[0]
         assert jac_eqn.invars[0].aval == jaxpr.in_avals[0]
-        diff_eqn_assertions(jac_eqn, jacobian_prim)
+        diff_eqn_assertions(
+            jac_eqn,
+            scalar_out=False,
+        )
```
Comment (dwierichs) [inline tests/capture/test_capture_diff.py]:
```suggestion
        diff_eqn_assertions(jac_eqn, scalar_out=False)
```

File: tests/capture/test_capture_diff.py [hunk 16]
```diff
@@ -597,13 +606,15 @@ def func_jax(x):
         assert len(jaxpr.eqns) == 3
 
         argnums = [argnums] if isinstance(argnums, int) else argnums
-        # Compute the flat argnums in order to determine the expected number of out tracers
+        # Compute the flat argnum in order to determine the expected number of out tracers
```
Comment (dwierichs) [inline tests/capture/test_capture_diff.py]: 🤔

File: doc/releases/changelog-dev.md (hunk 8 above)
Comment (albi3ro) [reply doc/releases/changelog-dev.md]: Sure 👍 I can go with `jacobian_p` instead.

File: pennylane/tape/plxpr_conversion.py [hunk 17]
```diff
@@ -184,17 +183,12 @@ def _(self, *wires, pauli_word="", postselect=None):
...
-@CollectOpsandMeas.register_primitive(grad_prim)
-def _grad_primitive(self, *invals, jaxpr, n_consts, **params):
-    raise NotImplementedError("CollectOpsandMeas cannot handle the grad primitive")
-
-
-# pylint: disable=unused-argument
 @CollectOpsandMeas.register_primitive(jacobian_prim)
 def _jacobian_primitive(self, *invals, jaxpr, n_consts, **params):
-    raise NotImplementedError("CollectOpsandMeas cannot handle the jacobian primitive")
+    raise NotImplementedError("CollectOpsandMeas cannot handle the grad primitive")
```
Comment (albi3ro) [inline pennylane/tape/plxpr_conversion.py]:
```suggestion
    raise NotImplementedError("CollectOpsandMeas cannot handle the jacobian primitive")
```

File: tests/capture/test_capture_diff.py (hunk 16 above)
Comment (albi3ro) [reply tests/capture/test_capture_diff.py]:
```suggestion
        # Compute the flat argnums in order to determine the expected number of out tracers
```

File: pennylane/capture/base_interpreter.py [hunk 18]
```diff
@@ -609,23 +608,12 @@ def handle_qnode(self, *invals, shots_len, qnode, device, execution_config, qfun
     )
 
 
-@PlxprInterpreter.register_primitive(grad_prim)
+@PlxprInterpreter.register_primitive(jacobian_prim)
 def handle_grad(self, *invals, jaxpr, n_consts, **params):
     """Handle the grad primitive."""
```
Comment (dwierichs) [inline pennylane/capture/base_interpreter.py]:
```suggestion
def handle_jacobian(self, *invals, jaxpr, n_consts, **params):
"""Handle the jacobian primitive."""
```

File: pennylane/_grad.py [hunk 19]
```diff
@@ -37,85 +39,59 @@
...
-@lru_cache
-def _get_jacobian_prim():
-    """Create a primitive for Jacobian computations.
-    This primitive is used when capturing ``qml.jacobian``.
-    """
-    if not has_jax:  # pragma: no cover
-        return None
-
-    jacobian_prim = capture.QmlPrimitive("jacobian")
+    jacobian_prim = capture.QmlPrimitive("grad")
```
Comment (dwierichs) [inline pennylane/_grad.py]: Am I confused? 🤔 
```suggestion
    jacobian_prim = capture.QmlPrimitive("jacobian")
```
